### PR TITLE
Addition of Migration Interceptor change

### PR DESCRIPTION
**Pull request by abhishekj720**
Added executeWorkflow method and logic for migration in new domain
- what to do when cron and child workflows
- what if they are not
- added versioning

### CODE REVIEW

File: src/main/java/com/uber/cadence/migration/MigrationInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+          try {
+            MigrationActivities.StartNewWorkflowExecutionResponse response =
+                activities.startWorkflowInNewDomain(
+                    new MigrationActivities.StartNewWorkflowRequest());
+            // TODO: add metrics and logging
+            throw new CancellationException(
+                "cancel due to migration:" + response.response.toString());
+          } catch (ActivityException e) {
+            // fallback if start workflow in new domain failed
+            next.executeWorkflow(workflowDefinition, input);
```
Comment (shijiesheng) [inline src/main/java/com/uber/cadence/migration/MigrationInterceptor.java:131]:
```suggestion
            return next.executeWorkflow(workflowDefinition, input);
```

File: src/main/java/com/uber/cadence/migration/MigrationInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+          }
+        }
+    }
+    return next.executeWorkflow(workflowDefinition, input);
```
Comment (shijiesheng) [inline src/main/java/com/uber/cadence/migration/MigrationInterceptor.java:135]: delete this line, let's conclude return in every switch branch
```suggestion
```

File: src/main/java/com/uber/cadence/migration/MigrationInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+                activities.startWorkflowInNewDomain(
+                    new MigrationActivities.StartNewWorkflowRequest());
+            // TODO: add metrics and logging
+            throw new CancellationException(
+                "cancel due to migration:" + response.response.toString());
+          } catch (ActivityException e) {
+            // fallback if start workflow in new domain failed
+            next.executeWorkflow(workflowDefinition, input);
+          }
+        }
```
Comment (shijiesheng) [inline src/main/java/com/uber/cadence/migration/MigrationInterceptor.java:133]:
```suggestion
            default:
                return next.executeWorkflow(workflowDefinition, input);
        }
```

File: src/main/java/com/uber/cadence/migration/MigrationInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                          .setName(
+                              input
+                                  .getWorkflowExecutionStartedEventAttributes()
+                                  .taskList
+                                  .getName()))
+                  .setInput(input.getInput())
+                  .setWorkflowType(new WorkflowType().setName(input.getWorkflowType().getName()))
+                  .setWorkflowIdReusePolicy(WorkflowIdReusePolicy.TerminateIfRunning)
+                  .setRetryPolicy(
+                      input.getWorkflowExecutionStartedEventAttributes().getRetryPolicy())
```
Comment (shijiesheng) [inline src/main/java/com/uber/cadence/migration/MigrationInterceptor.java:100]: use `startEventAttr`

File: src/main/java/com/uber/cadence/migration/MigrationInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  @Override
+  public byte[] executeWorkflow(
+      SyncWorkflowDefinition workflowDefinition, WorkflowExecuteInput input) {
+
+    WorkflowInfo workflowInfo = Workflow.getWorkflowInfo();
+    int version = getVersion(versionChangeID, Workflow.DEFAULT_VERSION, versionV1);
+    switch (version) {
+      case versionV1:
+        // Skip migration on non-cron and child workflows
+        if (input.getWorkflowExecutionStartedEventAttributes().cronSchedule == ""
```
Comment (shijiesheng) [inline src/main/java/com/uber/cadence/migration/MigrationInterceptor.java:71]:
```suggestion
        WorkflowExecutionStartedEventAttributes startEventAttr
              = input.getWorkflowExecutionStartedEventAttributes();
        if (startEventAttr.cronSchedule == "" || startEventAttr.getParentWorkflowExecution().getWorkflow() != "")
```

File: src/main/java/com/uber/cadence/migration/MigrationInterceptor.java (hunk 2 above)
Comment (abhishekj720) [reply src/main/java/com/uber/cadence/migration/MigrationInterceptor.java:135]: yeah, tried that. but even if we have default that returns the next, it shows missing return statement

**Pull request merged**